### ISSUE TITLE
Selecting element in typeahead does not trigger parsers

### ISSUE THREAD

**Issue by zvh421** [angular-ui/bootstrap#5283]
I'm not sure if this is intended or not, but my custom parser does not get triggered when selecting a value in a typeahead.

As you can see in plunker example, when I type individual characters the custom parser is triggered; but if I actually select a value from the list the parser is not triggered.

Is there a way I can make it so that my parser gets triggered both while typing and on selection?

http://plnkr.co/edit/l9G7QTfhjdq9NYhvxEJ9?p=preview

Thanks!

**Comment by wesleycho** [comment]
Unfortunately, what might be happening is that the typeahead is changing the model value - it should probably be calling $setViewValue on selection to trigger the model change. This might require a refactor to support this.

**Comment by zvh421** [comment]
Thanks for the answer; do you have any suggested workaround in the meantime? I see that there is an on-select callback but not sure how I would call $setViewValue from there (if appropriate)

**Comment by wesleycho** [comment]
Unfortunately, this may be something that needs to be fixed at the library level.

Can you explain your use case more in detail?

**Comment by zvh421** [comment]
I am attempting to convert an existing input field to a typeahead. The existing input field has two custom parsers. 

The input field is currently designed such that the user must enter an entity name followed by a command. As a contrived example let's say the entities are pets and the commands are things you can do to the pet. So the user would enter the name for the pet and then the operation like "Rex feed".

The first parser checks to make sure that the first word is a valid pet name; if it is, it then returns the command portion of the input minus the pet name ("feed"). If it is not a valid pet name, it returns an empty string.

The second parser receives the command minus the pet name ("feed") and then checks to make sure the command has not been "blacklisted", and if it has been, suggests an alternative in a nearby div and returns an empty value. If the command has not been blacklisted, it returns the input value "feed"; it is important that the value returned by the last parser is the command only and does not contain the pet name.

There may be a better UI for doing this, but I was hoping to not have to change it because users are already used to the current way of doing things.

Thanks so much for the help!

**Comment by RootMale** [comment]
+1 Its definetely should trigger setviewvalue or it brake angular forms functionality

**Comment by lafeuil** [comment]
I agree with @RootMale. The typeahead component brakes angular forms functionality. 

In the code, when a value is selected in dropdown menu, the [modelValue is setted](https://github.com/angular-ui/bootstrap/blob/master/src/typeahead/typeahead.js#L356) directly. 
I think, the viewValue and input element value should be defined instead :
```javascript
modelCtrl.$setViewValue(model, 'select');
element.val(parserResult.viewMapper(originalScope, locals));
```
The modelValue will be updated automatically with angular form.

Tell me if you agree with this and so, I will create a pull request.

**Comment by wesleycho** [comment]
The typeahead does need fixing with how it handles being part of a form - unfortunately I do not have the time to tackle such a task myself.

**Comment by lafeuil** [comment]
@wesleycho, you are right.
I tried to adapt typeahead to angular forms functionality. But I realized that is more complicated that I thought.
